flexmeasures/data/models/planning/storage.py

```diff
@@ -368,50 +386,345 @@ def build_device_soc_targets(
+    ########################
+
+    # 1) min >= soc_min
+    soc_min = (soc_min - soc_at_start) * timedelta(hours=1) / resolution
+    _constraints["soc_min(t)"] = soc_min
+    constraint_violations += validate_constraint(_constraints, "soc_min(t) <= min(t)")
+
+    # 2) max <= soc_max
+    soc_max = (soc_max - soc_at_start) * timedelta(hours=1) / resolution
+    _constraints["soc_max(t)"] = soc_max
+    constraint_violations += validate_constraint(_constraints, "max(t) <= soc_max(t)")
+
+    ########################################
+    # B. Validation in the same time frame #
+    ########################################
+
+    # 1) min <= max
+    constraint_violations += validate_constraint(_constraints, "min(t) <= max(t)")
+
+    # 2) min <= equals
+    constraint_violations += validate_constraint(_constraints, "min(t) <= equals(t)")
+
+    # 3) equals <= max
+    constraint_violations += validate_constraint(_constraints, "equals(t) <= max(t)")
+
+    ##########################################
+    # C. Validation in different time frames #
+    ##########################################
+
+    _constraints["factor_w_wh(t)"] = resolution / timedelta(hours=1)
+    _constraints["min(t-1)"] = prepend_serie(_constraints["min(t)"], soc_min)
+    _constraints["equals(t-1)"] = prepend_serie(_constraints["equals(t)"], soc_at_start)
+    _constraints["max(t-1)"] = prepend_serie(_constraints["max(t)"], soc_max)
+
+    # 1) equals(t) - equals(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "equals(t) - equals(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 2) derivative_min(t) <= equals(t) - equals(t-1)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= equals(t) - equals(t-1)"
+    )
+
+    # 3) min(t) - max(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "min(t) - max(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 4) max(t) - min(t-1) >= derivative_min(t)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= max(t) - min(t-1)"
+    )
+
+    # 5) equals(t) - max(t-1) <= derivative_max(t)
+    constraint_violations += validate_constraint(
+        _constraints, "equals(t) - max(t-1) <= derivative_max(t) * factor_w_wh(t)"
+    )
+
+    # 6) derivative_min(t) <= equals(t) - min(t-1)
+    constraint_violations += validate_constraint(
+        _constraints, "derivative_min(t) * factor_w_wh(t) <= equals(t) - min(t-1)"
+    )
+
+    return constraint_violations
+
+
+def get_pattern_match_word(word: str) -> str:
+    """Get a regex pattern to match a word
+
+    The conditions to delimit a word are:
+      - start of line
+      - whitespace
+      - end of line
+      - word boundary
+      - arithmetic operations
+
+    :return: regex expression
+    """
+
+    regex = r"(^|\s|$|\b|\+|\-|\*|/\|\\)"
```

Comment:
regex101.com suggests this contains an error:
```
/ An unescaped delimiter must be escaped; in most languages with a backslash (\)
```

It suggests escaping the `/`:
```
(^|\s|$|\b|\+|\-|\*|\/\|\\)
```

Maybe Python's `re` does the escape for us, I'm not sure.

Reply:
Great catch!